File: Data Structures/Singly Linked List/Java/SinglyLinkedList.java



public class SinglyLinkedList<T> {

    private int size;
    private Node<T> head;

    class Node<T> {
        private T data;
        private Node<T> nextNode;

        public Node(T data, Node<T> nextNode) {
            this.data = data;
            this.nextNode = nextNode;
        }

        public T getData() {
            return data;
        }

        public Node<T> getNextNode() {
            return nextNode;
        }

        public void setData(T data) {
            this.data = data;
        }

        public void setNextNode(Node<T> nextNode) {
            this.nextNode = nextNode;
        }
    }

    public boolean isEmpty() {
        return head == null;
    }

    public int getSize() {
        return size;
    }

    public void prepend(T data) {
        head = new Node<T>(data, head);
        size++;
    }

    public void append(T data) {
        if (head == null) {
            prepend(data);
            return;
        }
        Node<T> n = head;
        for (; n.getNextNode() != null ; n = n.getNextNode());
        n.setNextNode(new Node<T>(data, null));
        size++;
    }

    public T get(int index) {
        validateIndex(index);
        Node<T> n = head;
        int i = 0;
        for (; i < index; n = n.getNextNode(), i++);
        return n.getData();
    }

    public void set(int index, T data) {
        validateIndex(index);
        Node<T> n = head;
        int i = 0;
        for (; i < index; n = n.getNextNode(), i++);
        n.setData(data);
    }

    public int find(T data) {
        if (head == null) {
            return -1;
        }
        Node<T> n = head;
        for (int i = 0; n != null; n = n.getNextNode(), i++) {
            if (n.getData() != null && n.getData().equals(data)) {
                return i;
            }
        }
        return -1;
    }

    public void delete(int index) {
        validateIndex(index);
        Node<T> n = head;
        int i = 0;
        for (; i < index-1; n = n.getNextNode(), i++);
        n.setNextNode(n.getNextNode().getNextNode());
        size--;
    }

    public void add(int index, T data) {
        validateIndex(index);
        Node<T> n = head;
        int i = 0;
        for (; i < index-1; n = n.getNextNode(), i++);
        n.setNextNode(new Node<T>(data, n.getNextNode()));
        size++;
    }

    @Override
    public String toString(){
        if (isEmpty()) {
            return "size: 0\nhead -> null";
        }
        StringBuilder sb = new StringBuilder();
        sb.append("size: ");
        sb.append(size);
        sb.append("\nhead -> ");
        for (Node<T> n = head; n != null; n = n.getNextNode()) {
            sb.append(n.getData());
            sb.append(" -> ");
        }
        sb.append("null");
        return sb.toString();
    }

    private void validateIndex(int index) {
          if (index < 0 || index >= size) {
              throw new IndexOutOfBoundsException();
          }
    }

    public static void main(String []args) {
        SinglyLinkedList<Integer> list = new SinglyLinkedList<Integer>();
        System.out.println(list.toString());
        System.out.println("Operation: append(24)");
        list.append(24);
        System.out.println(list.toString());
        System.out.println("Operation: append(16)");
        list.append(16);
        System.out.println(list.toString());
        System.out.println("Operation: prepend(81)");
        list.prepend(81);
        System.out.println(list.toString());
        System.out.println("Operation: prepend(10)");
        list.prepend(10);
        System.out.println(list.toString());
        System.out.println("Operation: get(2)");
        System.out.println(list.get(2));
        System.out.println("Operation: add(2,75)");
        list.add(2,75);
        System.out.println(list.toString());
        System.out.println("Operation: set(2,53)");
        list.set(2,53);
        System.out.println(list.toString());
        System.out.println("Operation: find(99)");
        System.out.println(list.find(99));
        System.out.println("Operation: find(24)");
        System.out.println(list.find(24));
        System.out.println("Operation: list.delete(4)");
        list.delete(4);
        System.out.println(list.toString());
        System.out.println("Operation: list.delete(3)");
        list.delete(3);
        System.out.println(list.toString());
    }
}

File: Data Structures/Singly Linked List/Python/singly_linked_list.py
class Node:
	def __init__(self,data):
		self.__data=data
		self.__next=None

	def get_data(self):
		return self.__data

	def set_data(self,data):
		self.__data=data

	def get_next(self):
		return self.__next

	def set_next(self,next_node):
		self.__next=next_node

class LinkedList:
	def __init__(self):
		self.__head=None
		self.__tail=None

	def get_head(self):
		return self.__head

	def get_tail(self):
		return self.__tail

	def add(self,data):
		new_node=Node(data)
		if self.get_head() is None:
			self.__head=new_node
			self.__tail=new_node
		else:
			self.__tail.set_next(new_node)
			self.__tail=new_node

	def display(self):
		temp=self.get_head()
		while temp is not None:
			print(temp.get_data())
			temp=temp.get_next()

	def find_node(self,data):
		temp=self.get_head()
		while temp is not None:
			if temp.get_data()==data:
				return temp
			temp=temp.get_next()

	def insert(self,data,data_before):
		new_node=Node(data)
		if data_before is None:
			new_node.set_next(self.__head)
			self.__head=new_node
			if new_node.get_next() is None:
				self.__tail=new_node
		else:
			node_before=self.find_node(data_before)
			if node_before is None:
				raise IndexError(f"Node not found {data_before}")
			new_node.set_next(node_before.get_next())
			node_before.set_next(new_node)
			if new_node.get_next() is None:
				self.__tail=new_node		

	def delete(self,data):
		node=self.find_node(data)
		if node is None:
			raise ValueError("Node not found")
		if node is self.get_head():
			self.__head=node.get_next()
			node.set_next(None)
			if node is self.get_tail():
				self.__tail=None
		else:
			temp=self.get_head()
			node_before=None
			while temp is not node:
				node_before=temp
				temp=temp.get_next()
			node_before.set_next(node.get_next())
			if node is self.get_tail():
				self.__tail=node_before
				self.__tail.set_next(None)

if __name__=="__main__":
	size=int(input())
	l=list(map(int,input().split()))
	ll=LinkedList()
	for i in l:
		ll.add(i)
	items=int(input())
	for _ in range(items):
		ll.delete(ll.get_head().get_data())
	print("Linked List")
	temp=ll.get_head()
	while temp is not None:
		print(f"->{temp.get_data()}",end="")
		temp=temp.get_next()
	print()
